crates/core/src/transport/received_packet_tracker.rs

```diff
@@ -0,0 +1,187 @@
+use crate::transport::MessageId;
+use crate::util::{SystemTime, TimeSource};
+use std::collections::{HashMap, VecDeque};
+use std::mem;
+use std::time::{Duration, Instant};
+
+const RETAIN_TIME: Duration = Duration::from_secs(60);
+const MAX_PENDING_RECEIPTS: usize = 20;
+
+/// This struct is responsible for tracking received packets and deciding when to send receipts
+/// from/to a specific peer.
+///
+/// The caller must report when packets are received using the `report_received_packets` method.
+/// The caller must also call `get_receipts` periodically to check if any receipts need to be sent.
+///
+/// `get_receipts` should be called whenever a packet is sent.
+///
+/// `get_receipts` must **also** be called every `MAX_CONFIRMATION_DELAY` (50ms) and
+/// if the returned list is not empty, the list should be sent as receipts immediately in a noop.
+/// This may look something like this:
+///
+/// ```
+/// use std::thread::sleep;
+/// let mut received_packet_tracker = todo!();
+/// loop {
+///    let receipts = received_packet_tracker.get_receipts();
+///     if !receipts.is_empty() {
+///        // Send receipts in a noop message
+///     }
+///     sleep(MAX_CONFIRMATION_DELAY);
+///  }
+/// ```
+pub(super) struct ReceivedPacketTracker<T: TimeSource> {
+    pending_receipts: Vec<MessageId>,
+    message_id_time: VecDeque<(MessageId, Instant)>,
+    time_by_message_id: HashMap<MessageId, Instant>,
+    time_source: T,
+}
+
+impl ReceivedPacketTracker<SystemTime> {
+    pub(super) fn new() -> Self {
+        ReceivedPacketTracker {
+            pending_receipts: Vec::new(),
+            message_id_time: VecDeque::new(),
+            time_by_message_id: HashMap::new(),
+            time_source: SystemTime,
+        }
+    }
+}
+
+impl<T: TimeSource> ReceivedPacketTracker<T> {
+    pub(super) fn report_received_packets(&mut self, message_id: MessageId) -> ReportResult {
+        self.cleanup();
+        if self.time_by_message_id.contains_key(&message_id) {
+            ReportResult::AlreadyReceived
+        } else {
+            self.time_by_message_id
```

Comment:
can we batch this somehow? seems a bit wasteful doing a syscall to get time for each packet

Reply:
Agreed, the [coarsetime](https://github.com/jedisct1/rust-coarsetime) crate may be the solution.